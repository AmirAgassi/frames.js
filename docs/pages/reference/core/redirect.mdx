# redirect

`redirect()` must be used only in response to `post_redirect` button clicks. It accepts full URL as an argument.

```tsx
import { Button, createFrames, redirect } from "@frames.js/core";

const frames = createFrames();
const handleRequest = frames(async (ctx) => {
  if (ctx.clickedButton?.action === "post_redirect") {
    // when post_redirect button is clicked you must return a redirect response
    return redirect("https://example.com");
  }

  return {
    buttons: [<Button action="post_redirect">Click me</Button>],
  };
});
```
