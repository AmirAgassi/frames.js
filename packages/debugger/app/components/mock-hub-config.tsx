import React from "react";
import { type MockHubActionContext } from "../utils/mock-hub-utils";

export function MockHubConfig({
  hubContext,
  setHubContext,
}: {
  hubContext: Partial<MockHubActionContext>;
  setHubContext: React.Dispatch<
    React.SetStateAction<Partial<MockHubActionContext>>
  >;
}) {
  console.log(hubContext);
  const handleInputChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const target = e.target as HTMLInputElement; // Adjusting the type here for broad compatibility
    const value = target.type === "checkbox" ? target.checked : target.value;
    const name = target.name;

    if (target.type === "checkbox") {
      setHubContext((prev) => ({ ...prev, [name]: value }));
    }
  };

  return (
    <div className={"text-gray-500"}>
      <div className="flex flex-col gap-2">
        <label className="flex items-center gap-2">
          <input
            name="enabled"
            type="checkbox"
            checked={hubContext.enabled}
            onChange={handleInputChange}
          />{" "}
          Mock Hub
        </label>
        {hubContext.enabled && (
          <>
            <label className="flex items-center gap-2">
              <input
                name="requesterFollowsCaster"
                type="checkbox"
                checked={hubContext.requesterFollowsCaster}
                onChange={handleInputChange}
              />{" "}
              Requester follows caster
            </label>
            <label className="flex items-center gap-2">
              <input
                name="casterFollowsRequester"
                type="checkbox"
                checked={hubContext.casterFollowsRequester}
                onChange={handleInputChange}
              />{" "}
              Caster follows requester
            </label>
            <label className="flex items-center gap-2">
              <input
                name="likedCast"
                type="checkbox"
                checked={hubContext.likedCast}
                onChange={handleInputChange}
              />{" "}
              Requester liked cast
            </label>
            <label className="flex items-center gap-2">
              <input
                name="recastedCast"
                type="checkbox"
                checked={hubContext.recastedCast}
                onChange={handleInputChange}
              />{" "}
              Requester recasted cast
            </label>
          </>
        )}
      </div>
    </div>
  );
}
