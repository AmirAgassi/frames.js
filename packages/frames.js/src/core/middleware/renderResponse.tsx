import { ImageResponse } from "@vercel/og";
import { type Frame, getFrameFlattened, getFrameHtmlHead } from "../..";
import type { ButtonProps } from "../components";
import type { FrameDefinition, FramesMiddleware } from "../types";
import { generatePostButtonTargetURL, isFrameRedirect } from "../utils";

/**
 * This middleware is responsible for rendering the response
 *
 * If the accept header is set to application/json, it will return the metadata as JSON
 * so it is easy to parse it for metatags in existing applications.
 */
export function renderResponse(): FramesMiddleware<{}> {
  return async (context, next) => {
    const wantsJSON =
      context.request.headers.get("accept") === "application/json";

    const result = await next(context);

    if (result instanceof Response) {
      return result;
    }

    if (isFrameRedirect(result)) {
      return new Response(null, {
        status: result.status ?? 302,
        headers: {
          ...result.headers,
          Location: result.location.toString(),
        },
      });
    }

    const frame: Frame = {
      version: "vNext",
      postUrl: "",
      state: result.state ? JSON.stringify(result.state) : undefined,
      // @todo rendering image could be moved to its own middleware instead so users can use something different if they want to?
      // but that would mean that we need to specify middleware manually in any app since importing it here in default middleware
      // and disabling it, has no effect on final bundle size of app
      image:
        typeof result.image === "string"
          ? result.image
          : await renderImage(result.image, result.imageOptions),
      buttons: result.buttons?.map(
        (button, i): NonNullable<Frame["buttons"]>[number] => {
          if (!("type" in button && "props" in button)) {
            throw new Error("Invalid button provided");
          }

          if (i > 3) {
            throw new Error("Only 4 buttons are allowed");
          }

          const props = button.props as ButtonProps;

          switch (props.action) {
            case "link":
            case "mint":
              return {
                action: props.action,
                label: props.children,
                target: props.target,
              };
            case "post":
            case "post_redirect":
              return {
                action: props.action,
                label: props.children,
                target: generatePostButtonTargetURL({
                  buttonIndex: (i + 1) as 1 | 2 | 3 | 4,
                  buttonAction: props.action,
                  target: props.target,
                  currentURL: context.currentURL,
                  basePath: context.basePath,
                  state: props.state,
                }),
              };
            default:
              throw new Error("Unrecognized button action");
          }
        }
      ) as Frame["buttons"],
      inputText: result.textInput,
      imageAspectRatio: result.imageOptions?.aspectRatio ?? "1.91:1",
    };

    if (wantsJSON) {
      const flattened = getFrameFlattened(frame);

      return new Response(JSON.stringify(flattened), {
        status: result.status ?? 200,
        statusText: result.statusText ?? "OK",
        headers: {
          ...result.headers,
          "content-type": "application/json",
        },
      });
    }

    return new Response(getFrameHtmlHead(frame), {
      status: result.status ?? 200,
      statusText: result.statusText ?? "OK",
      headers: {
        ...result.headers,
        "content-type": "text/html",
      },
    });
  };
}

async function renderImage(
  element: React.ReactElement,
  options: FrameDefinition["imageOptions"]
): Promise<string> {
  const response = new ImageResponse(
    (
      <div
        style={{
          display: "flex", // Use flex layout
          flexDirection: "row", // Align items horizontally
          alignItems: "stretch", // Stretch items to fill the container height
          width: "100%",
          height: "100vh", // Full viewport height
          backgroundColor: "white",
        }}
      >
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            justifyContent: "center",
            alignItems: "center",
            lineHeight: 1.2,
            fontSize: 36,
            color: "black",
            flex: 1,
            overflow: "hidden",
          }}
        >
          {element}
        </div>
      </div>
    ),
    {
      ...(options?.aspectRatio === "1:1"
        ? {
            width: 1146,
            height: 1146,
          }
        : { width: 1146, height: 600 }),
      ...options,
    }
  );

  const buffer = await response.arrayBuffer();

  return `data:image/png;base64,${Buffer.from(buffer).toString("base64")}`;
}
