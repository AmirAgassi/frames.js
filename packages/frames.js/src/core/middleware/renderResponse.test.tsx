import { Button } from "../components";
import { redirect } from "../redirect";
import type { FramesContext } from "../types";
import { renderResponse } from "./renderResponse";

jest.mock("@vercel/og", () => {
  return {
    ImageResponse: class {
      arrayBuffer() {
        return new ArrayBuffer(10);
      }
    },
  };
});

describe("renderResponse middleware", () => {
  const render = renderResponse();
  const context: FramesContext = {
    basePath: "/",
    request: new Request("https://example.com"),
    currentURL: new URL("https://example.com"),
  };

  beforeEach(() => {
    context.request = new Request("https://example.com");
  });

  it("returns redirect Response if redirect is returned from handler", async () => {
    const result = await render(context, async () => {
      return redirect("https://example.com");
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(302);
    expect((result as Response).headers.get("location")).toBe(
      "https://example.com"
    );
  });

  it("allows custom status code and headers for redirect", async () => {
    const result = await render(context, async () => {
      return redirect("https://example.com", {
        status: 307,
        headers: { "X-Test": "test" },
      });
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(307);
    expect((result as Response).headers.get("location")).toBe(
      "https://example.com"
    );
    expect((result as Response).headers.get("x-test")).toBe("test");
  });

  it("allows to render a Response frame with custom status code and headers", async () => {
    const result = await render(context, async () => {
      return new Response("Hello", {
        status: 201,
        headers: { "X-Test": "test" },
      });
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(201);
    expect((result as Response).headers.get("x-test")).toBe("test");
  });

  it("allows to render a frame with image defined as react element", async () => {
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button action="post" state="customStateValue1">
            Click me 1
          </Button>,
          <Button action="post" state={{ a: true }} target="/a/b">
            Click me 2
          </Button>,
          <Button action="post" state={10} target="/test">
            Click me 3
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });

  it("allows to render a frame as json if Accept header is used", async () => {
    context.request = new Request("https://example.com", {
      headers: {
        Accept: "application/json",
      },
    });
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button action="post" state="customStateValue1">
            Click me 1
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    expect((result as Response).headers.get("Content-Type")).toBe(
      "application/json"
    );
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });

  it("properly resolves against basePath", async () => {
    context.basePath = "/prefixed";
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button action="post" state="customStateValue1" target="test">
            Click me 1
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });

  it("renders text input", async () => {
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        textInput: "My input",
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });
});
