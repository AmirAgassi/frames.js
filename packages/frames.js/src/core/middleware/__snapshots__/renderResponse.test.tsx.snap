// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderResponse middleware allows to render a frame as json if Accept header is used 1`] = `"{"fc:frame":"vNext","fc:frame:image":"data:image/png;base64,AAAAAAAAAAAAAA==","fc:frame:post_url":"","fc:frame:image:aspect_ratio":"1.91:1","fc:frame:button:1":"Click me 1","fc:frame:button:1:action":"post","fc:frame:button:1:target":"https://example.com/?__bi=1%3Ap&__bs=%22customStateValue1%22"}"`;

exports[`renderResponse middleware allows to render a frame with image defined as react element 1`] = `"<meta name="og:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame" content="vNext"/><meta name="fc:frame:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame:post_url" content=""/><meta name="fc:frame:image:aspect_ratio" content="1.91:1"/><meta name="fc:frame:button:1" content="Click me 1"/><meta name="fc:frame:button:1:action" content="post"/><meta name="fc:frame:button:1:target" content="https://example.com/?__bi=1%3Ap&__bs=%22customStateValue1%22"/><meta name="fc:frame:button:2" content="Click me 2"/><meta name="fc:frame:button:2:action" content="post"/><meta name="fc:frame:button:2:target" content="https://example.com/a/b?__bi=2%3Ap&__bs=%7B%22a%22%3Atrue%7D"/><meta name="fc:frame:button:3" content="Click me 3"/><meta name="fc:frame:button:3:action" content="post"/><meta name="fc:frame:button:3:target" content="https://example.com/test?__bi=3%3Ap&__bs=10"/>"`;

exports[`renderResponse middleware properly resolves against basePath 1`] = `"<meta name="og:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame" content="vNext"/><meta name="fc:frame:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame:post_url" content=""/><meta name="fc:frame:image:aspect_ratio" content="1.91:1"/><meta name="fc:frame:button:1" content="Click me 1"/><meta name="fc:frame:button:1:action" content="post"/><meta name="fc:frame:button:1:target" content="https://example.com/prefixed/test?__bi=1%3Ap&__bs=%22customStateValue1%22"/>"`;

exports[`renderResponse middleware renders text input 1`] = `"<meta name="og:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame" content="vNext"/><meta name="fc:frame:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame:post_url" content=""/><meta name="fc:frame:image:aspect_ratio" content="1.91:1"/><meta name="fc:frame:input:text" content="My input"/>"`;
